streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -173,33 +173,24 @@ public void reinitializeStateStoresForPartitions(final Collection<TopicPartition
         return partitionsAndOffsets;
     }
 
-    List<ConsumerRecord<byte[], byte[]>> updateStandbyStates(final TopicPartition storePartition,
-                                                             final List<ConsumerRecord<byte[], byte[]>> records) {
+    void updateStandbyStates(final TopicPartition storePartition,
```

Comment:
Maybe we can still just do one pass over the records in this function, while passing in stream time wrapped in a `Long` object in parameter, and update it inside this function. Then we do not need to do this refactoring any more.

Reply:
I'm not sure I follow...

I've defined "stream time" as the highest observed timestamp on records whose offsets are below the limit.

This is why I'm iterating twice: The first pass filters out events that are over the limit and determines the stream time. This result gets passed to this method, which does a second iteration to convert the records to key/value format and figure out the last offset.

I can collapse this into one iteration by making `updateStandbyStates` expect a `List<KeyValue>` and add a paramter for the last offset. It wouldn't be able to validate its input, so it would have to trust that all the input records are below the limit and that the lastOffset is correct.

Reply:
> so it would have to trust that all the input records are below the limit and that the lastOffset is correct

Compared to your comment below:

> I'm not sure... it would be a bug in our code if it happened.

Thus, it seem to be a redundant check. Maybe it's worth to pass in `List<KeyValue>` to simplify the code.

This also align with your comment above:

> Now, we only pass in records whose offsets are under the limit, so there are never any left over to return.